mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,123 @@
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Make a get instance type request to ec2 hosts using [ImdsClient].
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<(), Error>
+    where
+        F: FnOnce(Result<String, Error>) + 'static,
+    {
+        let callback_closure = Box::new(callback);
+        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_raw_ptr = Box::into_raw(callback_wrapper) as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // SAFETY: `get_resource_callback_fn_ptr` is a guaranteed non-null callback function pointer pass to CRT.
+        // `callback_raw_ptr` is a pointer to a senquence of valid instructions in memory the binding function will
+        // drop boxed callback instructions and pointer once the corresponding crt function finish the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr)
+                .ok_or_last_error()
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` points to the memory area `aws_imds_client`,
+        // crt assures the safy of revoking memory.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
+
+/// Callback trait for when finished getting ec2 instance type.
+type OnGetResource = Box<dyn FnOnce(Result<String, Error>)>;
+struct ImdsClientGetResourceCallback(OnGetResource);
+
+/// Rust binding for CRT's callback function `aws_imds_client_on_get_resource_callback_fn`.
+unsafe extern "C" fn imds_client_get_resource_callback(
+    resource: *const aws_byte_buf,
+    error_code: i32,
+    user_data: *mut libc::c_void,
+) {
+    // SAFETY: CRT invokes the callback only once given it has `FnOnce` trait, the execution of callback
+    // instructions(rust closure) is in unsafe zone, however, the borrow checker assures that instuctions
+    // will not result memory safety issue.
+    let callback = Box::from_raw(user_data as *mut ImdsClientGetResourceCallback).0;
+
+    let result = if 0 != error_code {
+        Err(error_code.into())
+    } else {
+        let resource = std::slice::from_raw_parts((*resource).buffer, (*resource).len).to_vec();
```

Comment:
Why is it safe to dereference the `resource` pointer? Deserves a safety comment.